### PR TITLE
Add test for append and repair with partition simulator

### PR DESCRIPTION
**Pull request by shino**
This PR adds EQC (statem) test for file operations and repair under
simulated network partition.

The main purpose is to confirm no dataloss, i.e. every chunk that has
been successfully written (ACK received) by append/write opration will
be read after partition heals.

---

Test sometimes fails, following cases should be investigated respectively.
1. `bad_return_value` from `machi_cr_client`. Probably just code bug.
   Output: https://gist.github.com/shino/6e917e1afc26a83cb01e
2. `machi_cr_client:append_chunk` responds ok but `{Offset, Length, Key}`
   is the same one before witten. Maybe test code bug, maybe around simulator,
   maybe actual code... don't know yet...
   Output: https://gist.github.com/shino/f38140fbe1c4da8e8fa4
   (full version, long. https://gist.githubusercontent.com/shino/335e09373da5b46bb2c8/raw/59a84638cd916940f3e67af2c6a4b15eb4e551a9/gistfile1.txt)
3. `read_chunk` for written chunk returns different binary from the one when
   it was written. No idea at the moment
   Output: https://gist.github.com/shino/011f3d3e3f4028093d58#file-gistfile1-txt-L10569-L10580

### CODE REVIEW

**Comment by jadeallenx** [comment]
This looks very cool, @shino! Looking forward to running it tomorrow. :cake:

**Comment by slfritchie** [comment]
Bug #36 is an important one to fix.  My guess is that it addresses Shino's description of number two and number three?  I have not looked at the first one yet ... and probably will not have time today, sorry!  (Need to finish Ricon-related stuff and pack for travel tomorrow.)

**Comment by shino** [comment]
Agree > #36 will fix 2. and 3.

For the first one `bad_return_value`, quick fix is at https://github.com/basho/machi/commit/903e9395c8a32af8ae7440ab8d0fb4d170c93cfd, just `reply` tag missing for `handle_call` return. I will create another PR with small eunit unless too difficult :sweat_smile:

File: src/machi_cr_client.erl [hunk 1]
```diff
@@ -923,10 +927,22 @@ update_proj(#state{proj=undefined}=S) ->
...
     Timeout = 2*1000,
     WTimeout = 2*Timeout,
-    Proxies = orddict:to_list(ProxiesDict),
+    SimName = proplists:get_value(simulator_self_name, Opts, cr_client),
+    ExcludedFLUs =
+        case proplists:get_value(use_partition_simulator, Opts, false) of
+            true ->
+                Members = proplists:get_value(simulator_members, Opts, []),
+                {Partitions, _Islands} = machi_partition_simulator:get(Members),
+                [B || {A, B} <- Partitions, A =:= SimName];
```
Comment (slfritchie) [inline src/machi_cr_client.erl]: This should be A =:= SimName orelse B =:= SimName, to be affected in either direction.

**Comment by slfritchie** [comment]
+1 from me (just one small change request) but I have been in a hurry today.  @mrallen1 and/or @kuenishi, could you please review also?

Also/otherwise, I'm thrilled that this is finding bugs even before the PR is merged.  `^_^`

**Comment by kuenishi** [comment]
I'll try a review on Wednesday unless Mark takes while we're in holiday.

**Comment by shino** [comment]
@slfritchie Thanks for review! I pushed the commit to address your comment.

> For the first one `bad_return_value`, quick fix is at 903e939, just reply tag missing for handle_call return. I will create another PR with small eunit unless too difficult

My expectation was wrong and somewhat deeper. Opened Separate issue #39.

**Comment by kuenishi** [comment]
I often see (not always) this error:

``` erlang
Reason:
  exception:
    exit({critical_error,
            {doublewrite_diff,
               {1084, 10,
                <<112, 114, 101, 94, 48, 50, 51, 51, 98, 98, 51, 54, 45, 101,
                  54, 100, 101, 45, 52, 53, 55, 97, 45, 98, 98, 50, 50, 45, 50,
                  102, 97, 52, 97, 52, 49, 51, 56, 51, 51, 52, 94, 49>>},
               {<<197, 51, 119, 192, 169, 161, 98, 56, 134, 187>>,
                <<238, 54, 176, 107, 52, 147, 92, 85, 36, 126>>}}})
      in machi_ap_repair_eqc:append/3 (test/machi_ap_repair_eqc.erl:147)
         eqc_statem:run_parallel_commands/2 (../src/eqc_statem.erl:1449)
         machi_ap_repair_eqc:-prop_repair_par/1-fun-2-/3 (test/machi_ap_repair_eqc.erl:260)
machi_ap_repair_eqc:89: prop_repair_par_test_...*failed*
in function machi_ap_repair_eqc:'-prop_repair_par_test_/0-fun-1-'/3 (test/machi_ap_repair_eqc.erl, line 92)
```

It pretty look like the 3rd issue but not sure.

File: test/machi_ap_repair_eqc.erl [hunk 2]
```diff
@@ -0,0 +1,568 @@
...
+
+%% append
+
+append_args(#state{cr_list=CRList}=S) ->
+    [elements(CRList), chunk(), S].
+
+append({_SimSelfName, C}, Bin, #state{verbose=V}=S) ->
+    Prefix = <<"pre">>,
+    Len = byte_size(Bin),
+    Res = (catch machi_cr_client:append_chunk(C, Prefix, Bin, {sec(1), sec(1)})),
```
Comment (kuenishi) [inline test/machi_ap_repair_eqc.erl:119]: ~~`append_chunk/4` accepts positive integers as timeout argument. I don't think this is properly handled by `machi_cr_client:timeout/1`.~~ I was wrong.

Comment (kuenishi) [inline test/machi_ap_repair_eqc.erl:119]: Note: with a modification to change `{sec(1), sec(1)}` to `sec(1)`, tests looks passing. While this kind of test fix is not right, but could be a hint for finding root cause.

**Comment by shino** [comment]
> I often see (not always) this error:

That is 2nd one. Please refer #36 , which includes temporary(?) patch.

File: test/machi_ap_repair_eqc.erl (hunk 2 above)
Comment (shino) [inline test/machi_ap_repair_eqc.erl:119]: Hmm... How is the number of tests after `sec(1)`? Almost the same as before?

Comment (kuenishi) [inline test/machi_ap_repair_eqc.erl:119]: Oh, with `EQC_TIMEOUT=600` it failed. Great

Comment (shino) [inline test/machi_ap_repair_eqc.erl:119]: In general, the test interval of 60 sec is very short for the implementation, only 5 or so tests are executed... Using `EQC_TIMEOUT` may help, e.g.

```
% EQC_VERBOSE=true EQC_TIMEOUT=600 rebar eunit skip_deps=true  \
         suites=machi_ap_repair_eqc tests=prop_repair_par_test_
```

File: test/machi_ap_repair_eqc.erl [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+
+    [ets:new(T, [set, public, named_table,
+                 {write_concurrency, true}, {read_concurrency, true}]) ||
+        T <- tabs()],
+    {ok, _} = application:ensure_all_started(machi),
+
+    SimSpec = {part_sim,
+               {machi_partition_simulator, start_link, [{0,0,0}, 0, 100]},
+               permanent, 500, worker, []},
+    {ok, _PSimPid} = supervisor:start_child(machi_sup, SimSpec),
```
Comment (kuenishi) [inline test/machi_ap_repair_eqc.erl:338]: Partition simulator must be stopped at every eqc_fsm setup.

**Comment by kuenishi** [comment]
+1 but please address one small comment before merge. Great job!

File: test/machi_ap_repair_eqc.erl (hunk 3 above)
Comment (shino) [inline test/machi_ap_repair_eqc.erl:338]: In `shutdown_hard()`, application `machi` is stopped, which includes partition simulator process.

**Pull request merged**